modules/javafx.graphics/src/main/java/com/sun/marlin/OffHeapArray.java

```diff
@@ -26,82 +26,164 @@
 package com.sun.marlin;
 
 import static com.sun.marlin.MarlinConst.LOG_UNSAFE_MALLOC;
-import java.lang.reflect.Field;
-import sun.misc.Unsafe;
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.ValueLayout;
+import java.nio.ByteOrder;
+
+// KCR: BEGIN DEBUG
+import java.util.concurrent.atomic.AtomicInteger;
+// KCR: END DEBUG
 
-/**
- *
- */
-// FIXME: We must replace the terminally deprecated sun.misc.Unsafe
-// memory access methods; see JDK-8334137
-@SuppressWarnings("removal")
 final class OffHeapArray  {
 
-    // unsafe reference
-    static final Unsafe UNSAFE;
+    private Arena arena;
+
     // size of int / float
     static final int SIZE_INT;
+    // FFM stuff
+    private static final ValueLayout.OfByte BYTE_LAYOUT = ValueLayout.JAVA_BYTE;
+    private static final ValueLayout.OfInt INT_LAYOUT = ValueLayout.JAVA_INT.withOrder(ByteOrder.BIG_ENDIAN);
 
     static {
-        try {
-            final Field field = Unsafe.class.getDeclaredField("theUnsafe");
-            field.setAccessible(true);
-            UNSAFE = (Unsafe) field.get(null);
-        } catch (Exception e) {
-            throw new InternalError("Unable to get sun.misc.Unsafe instance", e);
-        }
-
-        SIZE_INT = Unsafe.ARRAY_INT_INDEX_SCALE;
+        // KCR FIXME: get this from FFM
+        SIZE_INT = 4;
     }
 
     /* members */
-    long address;
-    long length;
-    int  used;
+    private MemorySegment segment;
+//    private long address;
+    private long length;
+    private int used;
+
+    // KCR: BEGIN DEBUG
+    private static final AtomicInteger instanceCount = new AtomicInteger(0);
+    private final int instance;
+    // KCR: END DEBUG
 
     OffHeapArray(final Object parent, final long len) {
+        // KCR: BEGIN DEBUG
+        instance = instanceCount.incrementAndGet();
+        if (LOG_UNSAFE_MALLOC) {
+            System.err.println("OffHeapArray::<init> : " +
+                    "instance = " + instance +
+                    ", thread = " + Thread.currentThread());
+        }
+        // KCR: END DEBUG
+
+        arena = Arena.ofShared();
+
         // note: may throw OOME:
-        this.address = UNSAFE.allocateMemory(len);
+        // KCR FIXME: Set a MemoryLayout
+        this.segment = arena.allocate(len);
```

Comment:
I'll choose a memory layout and also specify 4-byte alignment.

Reply:
Yes 4-bytes min or more like 16 if it helps the compiler to use avx instructions for fill / copy ?